qiskit/compiler/assemble.py

```diff
@@ -208,6 +209,14 @@ def _parse_common_args(backend, qobj_id, qobj_header, shots,
                    **qobj_header}
     qobj_header = QobjHeader(**{k: v for k, v in qobj_header.items() if v is not None})
 
+    max_shots = getattr(backend_config, 'max_shots', None)
+    if max_shots and max_shots < shots:
+        warnings.warn(
+            'Number of shots specified: %s exceeds max_shots property of the '
+            'backend: %s. Reducing shots to max_shots' % (shots, max_shots),
+            RuntimeWarning)
+        shots = max_shots
```

Comment:
I don't think we should automatically change this as then there is no way to turn this off? Should we either make validation optional or max_shots configurable?

Reply:
I agree it doesn't seem idea, and I thought about just making this raise an error. But the problem is when the default of 1024 is greater than the device max. Like when you run:

```
from qiskit import *
from qiskit.providers.aqt import AQT
aqt = AQT.enable_account('token')
backend = aqt.get_backend('aqt_qasm_simulator')
qc = QuantumCircuit(2, 2)
qc.h(0)
qc.cx(0, 1)
qc.measure([0,1], [0,1])
execute(qc, backend)
```

you'll get a 400 response from the api because `shots` is > 250. It felt weird to me to raise an exception when `shots` is not specified. The only other way I could think to workaround this is to change the behavior of the `shots` kwarg and make 1024 the default or `max_shots` (whichever is lower) and when you set `shots` as a kwarg it raises if you exceed `max_shots`.

Reply:
I think what you propose might work.

Reply:
At some point, a `validation.py` module will be justified. Is now the time?

Reply:
Ok, I've respun this in https://github.com/Qiskit/qiskit-terra/pull/3606/commits/0dc5ce84ee6e4d717b3bc4be381f09f66cac56be to make this a hard exception and to adjust the default to either be `min(1024, max_shots)`